pkg/kubelet/cri/streaming/portforward/httpstream.go

```diff
@@ -242,8 +243,10 @@ Loop:
 // function for the given stream pair.
 func (h *httpStreamHandler) portForward(p *httpStreamPair) {
 	ctx := context.Background()
-	defer p.dataStream.Close()
-	defer p.errorStream.Close()
+	resetStreams := false
+	defer func() {
+		closeStreamPair(p, resetStreams)
```

Comment:
Now this looks like `closeStreamPair` is not actually needed. Would it work if you just move its code here?

Reply:
fixed, PTAL

Reply:
This is the code you had previously. It looks ok to me, but I was thinking about getting rid of an extra function and keeping all reset/close logic in one place, something like this may be?
resetStreams = false
```
defer func() {
	if resetStreams {
		p.dataStream.Reset()
		p.errorStream.Reset()
		return
	}
	p.dataStream.Close()
	p.errorStream.Close()
}
```

Reply:
Sorry, I missed your point 😅. A fix has been made,


I remember you asked a question before [If you haven't asked, then here is a supplement]

> Why clear reset stream instead of close stream?

I tested the PR in Containerd [the main way is to debug spdy through `export DEBUG=1`] The strange thing is that close does not seem to close the entire stream. As a result, the client side will be blocked for a long time. After replacing it with reset, the problem will naturally disappear. Maybe it is a bug of spdy? Do you have an answer here?

Reply:
> As a result, the client side will be blocked for a long time. After replacing it with reset, the problem will naturally disappear. 

what does the client does when you reset the stream? does it exit clean?

Reply:
> Reset() closes both directions of the stream

Reply:
but why it hangs?
I don't have clear where this patch is needed @sxllwx , is this needed for containerd for k/k or for both?
I think we can add an ee2e test, is just to use a portforwrd to a pod with a listening server, stop the listening server and try to send traffic, then it will hit the EPIPE error because the forwarded stream will not have anything in the other side

Reply:
> does the client does when you reset the stream? does it exit clean?

yes~

Reply:
> but why it hangs? I don't have clear where this patch is needed @sxllwx , is this needed for containerd for k/k or for both? I think we can add an ee2e test, is just to use a portforwrd to a pod with a listening server, stop the listening server and try to send traffic, then it will hit the EPIPE error because the forwarded stream will not have anything in the other side

xref: https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1543873482

Reply:
I've added an image to be able to add e2e tests https://github.com/kubernetes/kubernetes/pull/117442#issuecomment-1535532821 
and @brianpursley put together an e2e test https://github.com/kubernetes/kubernetes/pull/117851

can we respect the commits and aggregate everything in one PR?